File: csharp/1106.parsing-a-boolean-expression.cs
/*
 * @lc app=leetcode id=1106 lang=csharp
 *
 * [1106] Parsing A Boolean Expression
 */

// @lc code=start
public class Solution {

    private static void FindEnd(string expression, int n, ref int i) {
        int acc = 0;
        while (acc != -1) {
            char c = expression[i++];
            if (c == '(') {
                ++acc;
            } else if (c == ')') {
                --acc;
            }
        }
    }


    private static bool ParseBoolExprRecurse(string expression, int n, ref int i) {
        if (expression[i] == 't') {
            ++i;
            return true;
        }
        if (expression[i] == 'f') {
            ++i;
            return false;
        }
        if (expression[i] == '!') {
            i += 2;
            bool b = ParseBoolExprRecurse(expression, n, ref i);
            ++i;
            return !b;
        }
        if (expression[i] == '&') {
            i += 2;
            bool b = true;
            while (i < n) {
                b &= ParseBoolExprRecurse(expression, n, ref i);
                if (!b) {
                    FindEnd(expression, n, ref i);
                    break;
                }
                if (expression[i] == ')') {
                    ++i;
                    break;
                }
                ++i;
            }
            return b;
        }
        if (expression[i] == '|') {
            i += 2;
            bool b = false;
            while (i < n) {
                b |= ParseBoolExprRecurse(expression, n, ref i);
                if (b) {
                    FindEnd(expression, n, ref i);
                    break;
                }
                if (expression[i] == ')') {
                    ++i;
                    break;
                }
                ++i;
            }
            return b;
        }
        return false;
    }

    public bool ParseBoolExpr(string expression) {
        int i = 0;
        return ParseBoolExprRecurse(expression, expression.Length, ref i);
    }
}
// @lc code=end

File: java/1106.parsing-a-boolean-expression.java
/*
 * @lc app=leetcode id=1106 lang=java
 *
 * [1106] Parsing A Boolean Expression
 */

import java.util.Stack;

// @lc code=start
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> stack = new Stack<>();
        int n = expression.length();
        for (int i = 0; i < n; ++i) {
            char c = expression.charAt(i);
            if (c == '(' || c == ',') {
                continue;
            }
            if (c != ')') {
                stack.push(c);
                continue;
            }
            boolean and = true, or = false, last = false;
            char t;
            do {
                t = stack.pop();
                if (t == 't') {
                    or = true;
                    last = true;
                } else if (t == 'f') {
                    and = false;
                    last = false;
                } else {
                    break;
                }
            } while (true);
            if (t == '!') {
                stack.push(last ? 'f' : 't');
            } else if (t == '|') {
                stack.push(or ? 't' : 'f');
            } else {
                stack.push(and ? 't' : 'f');
            }
        }
        return stack.pop() == 't';
    }
}
// @lc code=end

